src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributes.cs

```diff
@@ -0,0 +1,199 @@
+    /// </typeparam>
+    public class ReadOnlyTagHelperAttributes<TAttribute> : IReadOnlyList<TAttribute>
+        where TAttribute : IReadOnlyTagHelperAttribute
+    {
+        /// <summary>
+        /// Instantiates a new instance of the <see cref="ReadOnlyTagHelperAttributes{TAttribute}"/> with an empty
+        /// collection.
+        /// </summary>
+        protected ReadOnlyTagHelperAttributes()
+        {
+            Attributes = new List<TAttribute>();
+        }
+
+        /// <summary>
+        /// Instantiates a new instance of the <see cref="ReadOnlyTagHelperAttributes{TAttribute}"/> with the specified
+        /// <paramref name="attributes"/>.
+        /// </summary>
+        /// <param name="attributes">The collection to wrap.</param>
+        public ReadOnlyTagHelperAttributes([NotNull] IEnumerable<TAttribute> attributes)
+        {
+            Attributes = new List<TAttribute>(attributes);
+        }
+
+        /// <summary>
+        /// The underlying collection of <typeparamref name="TAttribute"/>s.
+        /// </summary>
+        /// <remarks>Modifying this <see cref="List{TAttribute}"/>s affects what is exposed by this
+        /// <see cref="ReadOnlyTagHelperAttributes{TAttribute}"/> instance.</remarks>
+        protected List<TAttribute> Attributes { get; }
+
+        /// <inheritdoc />
+        public TAttribute this[int index]
+        {
+            get
+            {
+                return Attributes[index];
+            }
+        }
+
+        /// <summary>
+        /// Gets the first <typeparamref name="TAttribute"/> whose <typeparamref name="TAttribute.Name"/> matches the
+        /// specified <paramref name="name"/>.
+        /// </summary>
+        /// <param name="name">
+        /// The <typeparamref name="TAttribute.Name"/> of the <typeparamref name="TAttribute"/> to get.
+        /// </param>
+        /// <returns>The first <typeparamref name="TAttribute"/> whose <typeparamref name="TAttribute.Name"/> matches
+        /// the specified <paramref name="name"/>.
+        /// </returns>
+        /// <remarks><paramref name="name"/> is compared case-insensitively.</remarks>
+        public TAttribute this[[NotNull] string name]
+        {
+            get
+            {
+                return Attributes.FirstOrDefault(attribute => NameEquals(name, attribute));
+            }
+        }
+
+        /// <inheritdoc />
+        public int Count
+        {
+            get
+            {
+                return Attributes.Count;
+            }
+        }
+
+        /// <inheritdoc />
+        public bool IsReadOnly
+        {
+            get
+            {
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Determines whether a <typeparamref name="TAttribute"/> with the same <typeparamref name="TAttribute.Name"/>
+        /// and <typeparamref name="TAttribute.Value"/> exists in the collection.
```

Comment:
suggest just `Determines whether a <typeparamref name="TAttribute"/> matching <paramref name="item"/> exists in the collection.` otherwise these (oft-repeated) comments will have to change whenever `IReadOnlyTagHelperAttribute` or the comparison changes.